packages/app-builder-lib/src/publish/KeygenPublisher.ts

```diff
@@ -36,78 +107,143 @@ export class KeygenPublisher extends HttpPublisher {
         Accept: "application/vnd.api+json",
-        "Content-Length": dataLength,
-        "Keygen-Version": "1.0",
+        "Keygen-Version": "1.1",
+        Prefer: "no-redirect",
+      },
+    }
+
+    const data: RecursivePartial<KeygenArtifact> = {
+      type: "artifacts",
+      attributes: {
+        filename: fileName,
+        filetype: getCompleteExtname(fileName),
+        filesize: dataLength,
+        platform: this.info.platform,
+      },
+      relationships: {
+        release: {
+          data: {
+            type: "releases",
+            id: releaseId,
+          },
+        },
       },
     }
-    await httpExecutor.doApiRequest(configureRequestOptions(upload, this.auth, "PUT"), this.context.cancellationToken, requestProcessor)
+
+    log.debug({ data: JSON.stringify(data) }, "Keygen create artifact")
+
+    return parseJson(httpExecutor.request(configureRequestOptions(upload, this.auth, "POST"), this.context.cancellationToken, { data }))
   }
 
-  private async upsertRelease(fileName: string, dataLength: number): Promise<{ data: any; errors: any }> {
+  private async getOrCreateRelease(): Promise<{ data?: KeygenRelease; errors?: KeygenError[] }> {
+    try {
+      return await this.getRelease()
+    } catch (e) {
+      if (e.statusCode === 404) {
+        return this.createRelease()
+      }
+
+      throw e
+    }
+  }
+
+  private async getRelease(): Promise<{ data?: KeygenRelease; errors?: KeygenError[] }> {
     const req: RequestOptions = {
       hostname: this.hostname,
-      method: "PUT",
-      path: this.basePath,
+      path: `${this.basePath}/releases/${this.version}`,
+      headers: {
+        Accept: "application/vnd.api+json",
+        "Keygen-Version": "1.1",
+      },
+    }
+
+    return parseJson(httpExecutor.request(configureRequestOptions(req, this.auth, "GET"), this.context.cancellationToken, null))
+  }
+
+  private async createRelease(): Promise<{ data?: KeygenRelease; errors?: KeygenError[] }> {
+    const req: RequestOptions = {
+      hostname: this.hostname,
+      path: `${this.basePath}/releases`,
       headers: {
         "Content-Type": "application/vnd.api+json",
         Accept: "application/vnd.api+json",
-        "Keygen-Version": "1.0",
+        "Keygen-Version": "1.1",
       },
     }
-    const data = {
-      data: {
-        type: "release",
-        attributes: {
-          filename: fileName,
-          filetype: getCompleteExtname(fileName),
-          filesize: dataLength,
-          version: this.version,
-          platform: this.info.platform,
-          channel: this.info.channel || "stable",
-        },
-        relationships: {
-          product: {
-            data: {
-              type: "product",
-              id: this.info.product,
-            },
+
+    const data: RecursivePartial<KeygenRelease> = {
+      type: "releases",
+      attributes: {
+        version: this.version,
+        channel: this.info.channel || "stable",
+        status: "PUBLISHED",
```

Comment:
I'd prefer to start with a `DRAFT` status for the release, and publish only after all artifacts are uploaded, but I couldn't find a hook to accomplish that (i.e. letting the publisher perform an additional request once all artifacts are uploaded). LMK if one exists and I just happened to miss it.

Reply:
Hmmm, I don't see any in the code for that functionality, but it totally seems like a valid reason to have one added :) 
Seems like an `onArtifactUploaded` could be added here? Simply need to add a task (i.e. Promise) to the `taskManager` after the `scheduleUpload`
https://github.com/electron-userland/electron-builder/blob/03cc9b96e0be07ef3450e3992eafcb7fe903a853/packages/app-builder-lib/src/publish/PublishManager.ts#L123-L133

From my quick look, an `onAllArtifactsUploaded` hook could probably be added here:
https://github.com/electron-userland/electron-builder/blob/03cc9b96e0be07ef3450e3992eafcb7fe903a853/packages/app-builder-lib/src/publish/PublishManager.ts#L230-L241

Reply:
If only adding `onAllArtifactsUploaded`, which is probably all we need for now, the best place might just be adding it here:
https://github.com/electron-userland/electron-builder/blob/03cc9b96e0be07ef3450e3992eafcb7fe903a853/packages/app-builder-lib/src/index.ts#L113

Reply:
I was thinking about doing that in a separate follow-up PR, and then updating the Keygen publisher while I'm there. Is that cool? As-is, I think the current publisher works even if it does skip the draft stage.

Reply:
Works for me 😄

Reply:
Yeah, it's ready for review but I wanted to wait until #6920 is out (which includes #6909) before we consider merging it, so I'm keeping the PR in draft mode. Detailed more of the reasoning there in the PR description.

LMK if you want me to move it out of a draft state. 👍

Reply:
Published 23.2.0 + electron-updater@5.0.6

Reply:
Awesome. Marked this as ready for review. 👍